data/account/participationRegistry.go

```diff
@@ -332,61 +415,176 @@ func (db *participationDB) GetAll() (records []ParticipationRecord, err error) {
+				}
+
+				copy := record
+				updated[record.ParticipationID] = copy
+			}
 		}
 
-		// update id
-		_, err = tx.Exec(setRegistered, on, record.LastValid, id[:])
+		// Mark registered.
+		recordToRegister.RegisteredFirst = on
+		recordToRegister.RegisteredLast = recordToRegister.LastValid
+
+		err := db.updateRollingFields(ctx, tx, recordToRegister)
 		if err != nil {
-			return fmt.Errorf("unable to update registered key: %w", err)
+			return fmt.Errorf("unable to registering key with id (%s): %w", id, err)
 		}
+		updated[recordToRegister.ParticipationID] = recordToRegister
 
 		return nil
 	})
-}
 
-func (db *participationDB) Record(account basics.Address, round basics.Round, participationAction ParticipationAction) error {
-	field, ok := participationActionFields[participationAction]
-	if !ok {
-		return ErrUnknownParticipationAction
+	// Update cache
+	if err == nil {
+		db.mutex.Lock()
+		defer db.mutex.Unlock()
+
+		for id, record := range updated {
+			delete(db.dirty, id)
+			db.cache[id] = record
+		}
 	}
 
-	query := fmt.Sprintf(updateRollingFieldX, field)
+	return err
+}
 
-	return db.store.Wdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
-		r, err := tx.Exec(query, round, account[:])
-		if err != nil {
-			return err
+func (db *participationDB) Record(account basics.Address, round basics.Round, participationAction ParticipationAction) error {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+
+	matches := make([]ParticipationID, 0)
+	// At most one id should be updated.
+	for _, record := range db.cache {
+		if record.Account == account && record.FirstValid <= round && round <= record.LastValid {
+			matches = append(matches, record.ParticipationID)
 		}
+	}
 
-		// If multiple rows were changed there is probably a problem with Register.
-		rowsEffected, err := r.RowsAffected()
-		if err != nil {
-			return err
-		} else if rowsEffected > 1 {
-			return fmt.Errorf("too many rows effected: %d", rowsEffected)
+	// Good case, one key found.
+	if len(matches) == 1 {
+		record := db.cache[matches[0]]
+
+		switch participationAction {
+		case Vote:
+			record.LastVote = round
+		case BlockProposal:
+			record.LastBlockProposal = round
+		case CompactCertificate:
+			record.LastCompactCertificate = round
+		default:
+			return ErrUnknownParticipationAction
 		}
 
+		db.dirty[record.ParticipationID] = struct{}{}
+		copy := record
+		db.cache[record.ParticipationID] = copy
+		return nil
+	}
+
+	// This probably means there is a bug in the key participation registry Register implementation.
+	if len(matches) > 1 {
+		return ErrMultipleValidKeys
+	}
+
+	// This indicates the participation registry is not synchronized with agreement.
+	return ErrActiveKeyNotFound
+}
+
+func (db *participationDB) Flush() error {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+
+	if len(db.dirty) == 0 {
+		return nil
+	}
+
+	return db.store.Wdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
```

Comment:
Refactor to avoid creating the atomic transaction unless you're truely hoing to flush.

Reply:
This is the "Flush()" function which would be called between each round. Is there a way to do this without an Atomic transaction?

Reply:
no, you would still need the Atomic operation, but if you'll look into what you're doing there, you'll see that in some cases, no updates would be made.
What I'm asking is to figure this out before making the Atomic call.

Reply:
That check is happening with an early return above:
```
	if len(db.dirty) == 0 {
		return nil
	}
```

Reply:
Yes, but I was referring to the `ErrParticipationIDNotFound` case, which should be ( in my view ), tested early.
but thinking about that - if we have 10 part keys, and one of them would generate the unexpected `ErrParticipationIDNotFound` error - we still want to flush all the rest. So, summing this up:
- the method should update all the keys it can.
- add the log to the participationDB so that we can log the error right away : we have the potential of having multiple errors during a single Flush() invocation.
- return an error only if the caller can do something useful with it.